File: E-Commerce App Project (Tabbed)/Controllers/AccountLoginViewController.m
//
//  AccountLoginViewController.m
//  E-Commerce App Project (Tabbed)
//
//  Created by Rony Banik on 15/11/18.
//  Copyright © 2018 Rony Banik. All rights reserved.
//

#import "AccountLoginViewController.h"
#import "AccountOperations.h"
#import "Validation.h"

@interface AccountLoginViewController (){
    AccountOperations* accountOperationsObj;
    Validation *validatorObj;
}

@end

@implementation AccountLoginViewController
@synthesize loginFormEmail;
@synthesize loginFormPassword;
@synthesize validationStatusLabel;

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    accountOperationsObj = [[AccountOperations alloc]init];
    validatorObj = [[Validation alloc]init];
    [loginFormEmail ajw_attachValidator:[validatorObj emailValidator:validationStatusLabel]];
    
    [loginFormPassword ajw_attachValidator:[validatorObj requiredMinLengthValidator:@"Insert Your Password!" IntegerforMinLength:1 minLengthErrorMessage:@"It can't be just nothing!" withLabelForValidationRules:validationStatusLabel]];
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/
-(void)LoginAction:(id)sender{
    NSString *encryptedPassword = [accountOperationsObj sha1:loginFormPassword.text];
    
    NSString *secureKeyForServerAccess = @"sdfsdfsd38792F423F4528482B4D6250655368566D597133743677397A24432646294A40kjsdhfkjsdhf";
    NSDictionary *dataToSend = [NSDictionary dictionaryWithObjectsAndKeys:
                                secureKeyForServerAccess,@"secureKeyForServerAccess_Enabled",
                                @"CHECK_USER_LOGIN",@"actionRequest",loginFormEmail.text,@"email",
                                encryptedPassword,@"password",
                                nil];
        [accountOperationsObj sendRequestToServer:dataToSend callback:^(NSError *error, BOOL success, NSString *customErrorMessage){
            if (success) {
                // Check if we're in offline mode
                if ([customErrorMessage hasPrefix:@"Offline Mode:"]) {
                    [sender setTitle:@"Demo Mode" forState:UIControlStateNormal];
                    [sender setBackgroundColor:[UIColor orangeColor]];
                    [accountOperationsObj showOfflineNotification:self];
                    // Still navigate to home after showing notification
                    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                        [self.navigationController popToRootViewControllerAnimated:YES];
                    });
                } else if([customErrorMessage isEqualToString:@"Login Successful"]){
                    [sender setTitle:@"Logged In" forState:UIControlStateNormal];
                    [sender setBackgroundColor:[UIColor greenColor]];
                    [self.navigationController popToRootViewControllerAnimated:YES];
                }else{
                    [self generalAlerts:@"Alert!" withMessage:customErrorMessage withYesActionTitle:@"Cancel Log In!" withNoActionTitle:@"Back to Login!"];
                }
            }else{
                [self generalAlerts:@"Not Logged In!" withMessage:@"Something went wrong please try again!" withYesActionTitle:@"Cancel Log In" withNoActionTitle:@"Ok!"];
                [sender setTitle:@"Try Again" forState:UIControlStateNormal];
            }
        }];
}
- (IBAction)loginFormSignInButton:(id)sender {
    
    if ([loginFormEmail hasText]&&[loginFormPassword hasText]) {
        if ([accountOperationsObj validateEmailAccount:loginFormEmail.text]) {
            [self LoginAction:sender];
        }else{
            [self generalAlerts:@"Invalid Email!" withMessage:@"Please Insert a valid email address" withYesActionTitle:@"Cancel Log In" withNoActionTitle:@"Ok!"];
        }
    }else{
        [self generalAlerts:@"Empty Form!" withMessage:@"Give your Email and Password" withYesActionTitle:@"Cancel Log In" withNoActionTitle:@"Ok!"];
    }
}

-(void)generalAlerts:(NSString *)alertTitle
         withMessage:(NSString *)alertMessage
  withYesActionTitle:(NSString *)yesActionTitle
   withNoActionTitle:(NSString *)noActionTitle{
    
    UIAlertController *generalAlert;
    generalAlert = [UIAlertController
                    alertControllerWithTitle:alertTitle
                    message:alertMessage
                    preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction* noAction = [UIAlertAction
                               actionWithTitle:yesActionTitle
                               style:UIAlertActionStyleDefault
                               handler:^(UIAlertAction * action) {
                                   //Handle your yes please button action here
                                   [self.navigationController popToRootViewControllerAnimated:YES];
                               }];
    UIAlertAction* yesAction = [UIAlertAction
                                actionWithTitle:noActionTitle
                                style:UIAlertActionStyleDefault
                                handler:^(UIAlertAction * action) {
                                    
                                }];
    [generalAlert addAction:yesAction];
    [generalAlert addAction:noAction];
    [self presentViewController:generalAlert animated:YES completion:nil];
}
@end

File: E-Commerce App Project (Tabbed)/Controllers/AccountLoginViewController.swift
//
//  AccountLoginViewController.swift
//  E-Commerce App Project (Tabbed)
//
//  Converted to Swift
//

import UIKit

class AccountLoginViewController: UIViewController {
    
    @IBOutlet var loginFormEmail: UITextField!
    @IBOutlet var loginFormPassword: UITextField!
    @IBOutlet var validationStatusLabel: UILabel!
    
    private var accountOperationsObj: AccountOperations!
    private var validatorObj: Validation!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        accountOperationsObj = AccountOperations()
        validatorObj = Validation()
        
        loginFormEmail.ajw_attachValidator(validatorObj.emailValidator(validationStatusLabel))
        loginFormPassword.ajw_attachValidator(validatorObj.requiredMinLengthValidator("Insert Your Password!", integerforMinLength: 1, minLengthErrorMessage: "It can't be just nothing!", withLabelForValidationRules: validationStatusLabel))
    }
    
    func loginAction(_ sender: UIButton) {
        let encryptedPassword = accountOperationsObj.sha1(loginFormPassword.text ?? "")
        
        let secureKeyForServerAccess = "sdfsdfsd38792F423F4528482B4D6250655368566D597133743677397A24432646294A40kjsdhfkjsdhf"
        let dataToSend: [String: Any] = [
            "secureKeyForServerAccess_Enabled": secureKeyForServerAccess,
            "actionRequest": "CHECK_USER_LOGIN",
            "email": loginFormEmail.text ?? "",
            "password": encryptedPassword
        ]
        
        accountOperationsObj.sendRequest(toServer: dataToSend) { [weak self] error, success, customErrorMessage in
            guard let self = self else { return }
            
            if success {
                if let message = customErrorMessage, message.hasPrefix("Offline Mode:") {
                    sender.setTitle("Demo Mode", for: .normal)
                    sender.backgroundColor = .orange
                    self.accountOperationsObj.showOfflineNotification(self)
                    DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
                        self.navigationController?.popToRootViewController(animated: true)
                    }
                } else if customErrorMessage == "Login Successful" {
                    sender.setTitle("Logged In", for: .normal)
                    sender.backgroundColor = .green
                    self.navigationController?.popToRootViewController(animated: true)
                } else {
                    self.generalAlerts("Alert!", withMessage: customErrorMessage ?? "", withYesActionTitle: "Cancel Log In!", withNoActionTitle: "Back to Login!")
                }
            } else {
                self.generalAlerts("Not Logged In!", withMessage: "Something went wrong please try again!", withYesActionTitle: "Cancel Log In", withNoActionTitle: "Ok!")
                sender.setTitle("Try Again", for: .normal)
            }
        }
    }
    
    @IBAction func loginFormSignInButton(_ sender: UIButton) {
        if loginFormEmail.hasText && loginFormPassword.hasText {
            if accountOperationsObj.validateEmailAccount(loginFormEmail.text ?? "") {
                loginAction(sender)
            } else {
                generalAlerts("Invalid Email!", withMessage: "Please Insert a valid email address", withYesActionTitle: "Cancel Log In", withNoActionTitle: "Ok!")
            }
        } else {
            generalAlerts("Empty Form!", withMessage: "Give your Email and Password", withYesActionTitle: "Cancel Log In", withNoActionTitle: "Ok!")
        }
    }
    
    func generalAlerts(_ alertTitle: String, withMessage alertMessage: String, withYesActionTitle yesActionTitle: String, withNoActionTitle noActionTitle: String) {
        let generalAlert = UIAlertController(
            title: alertTitle,
            message: alertMessage,
            preferredStyle: .alert
        )
        
        let noAction = UIAlertAction(title: yesActionTitle, style: .default) { [weak self] _ in
            self?.navigationController?.popToRootViewController(animated: true)
        }
        
        let yesAction = UIAlertAction(title: noActionTitle, style: .default, handler: nil)
        
        generalAlert.addAction(yesAction)
        generalAlert.addAction(noAction)
        present(generalAlert, animated: true, completion: nil)
    }
}
